text/0000-item-level-scopes.md

```diff
@@ -0,0 +1,181 @@
+- Feature Name: Item-Level Scopes
+- Start Date: 2018-03-29
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Allow Items to be grouped into "Item-level scopes".
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes one wishes to import a handful of functions/data structures from
+other modules, but only under some compile-time condition. For example, we
+might like to conditionally import some stuff only if we are in debug mode like
+this:
+
+```rust
+#[cfg(debug_assertions)]
+use a;
+#[cfg(debug_assertions)]
+use b;
+#[cfg(debug_assertions)]
+use c;
+#[cfg(debug_assertions)]
+use d
+```
+
+The above example is quite verbose because the `#[cfg...]` line must be
+repeated for each `use` line. This RFC proposes that the following should be
+possible:
+
+```rust
+#[cfg(debug_assertions)] {
+    use a;
+    use b;
+    use c;
+    use d;
+}
+```
+
+The latter is more concise and thus easier to read and write.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+An Item-level scope is allowed wherever an Item may appear. Item-level scopes
+are useful for (e.g.) compile-time guarding a collection of `use` statements.
+
+The following:
+
+```rust
+#[cfg(debug_assertions)]
+use a;
+#[cfg(debug_assertions)]
+use b;
+#[cfg(debug_assertions)]
+use c;
+#[cfg(debug_assertions)]
+use d
+```
+
+Can be written more concisely as:
+
+```rust
+#[cfg(debug_assertions)] {
+    use a;
+    use b;
+    use c;
+    use d;
+}
+```
```

Comment:
This example suggests that there won't be any actual scoping happening, it appears to be expecting the "scoped" `use` statements will be exported to the parent scope.

For example, would this compile or be an error:

```rust
{
    use a;
}

fn foo() {
    a::foo();
}
```

If it is actually introducing scopes then I think an actual use of the `use`s should be added to the examples to make them more like real code.

Reply:
Agreed.

Reply:
Actually wait. I think the idea is that things inside the `{}` are visible at the top-level.

So this should be possible:
```
#[cfg(target_os = "linux")] {
    use a, b, c, d;
}
...
#[cfg(target_os = "linux")]
fn do_linux_things() {
    a::blah();
}
```

Reply:
But I do see that there is an inconsistency. In regular Rust code, uses are scoped. So this doesn't work:

```
fn f () {
    let _ = Path::new("lalala");
}

fn main() {
    use std::path::Path;
    f();
}
```

Reply:
The only other idea I have (which is less pretty) is:
```
#[cfg(...)]
parent! {
    ...
}
```

Where the purpose of `parent!` is to make it explicit that the new scope is "injected" into the parent scope. But then that raises questions about whether such an interface is general:

```
fn f() {
    {
        parent! {
            let x = 1;
        }
        println!("{}", x);
    }
}
```

Should work?!

I don't see any need for such a use-case, so I'd favour not going down that path, but instead documenting that item-level scopes re-export to the top-level (module) scope. Full stop.

Reply:
By phrasing I was mostly referring to the documentation/naming of the feature. Mostly about removing the word "scope" from the name since that is directly related to scoping. One alternative would be to focus on "block" instead, maybe "conditional item blocks"?

Reply:
By phrasing I was mostly referring to the documentation/naming of the feature. Mostly about removing the word "scope" from the name since that visibility and scope are very tightly coupled terms. One alternative would be to focus on "block" instead, maybe "conditional item blocks"? Blocks are associated with visibility, but less so than scope.